x/ccv/provider/keeper/distribution.go

```diff
@@ -67,142 +66,265 @@ func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenom
-				"fail to allocate ICS rewards to community pool",
-				"consumerId", consumerId,
-				"chainId", chainId,
-				"error", err.Error(),
-			)
-			continue
-		}
+	// compute rewards for validators
+	consumerRewards := alloc.Rewards
+	voteMultiplier := math.LegacyOneDec().Sub(communityTax)
+	validatorsRewards := consumerRewards.MulDecTruncate(voteMultiplier)
 
-		// set consumer allocations to the remaining rewards decimals
-		alloc.Rewards = validatorsRewardsChange.Add(remainingChanges...)
-		k.SetConsumerRewardsAllocation(ctx, consumerId, alloc)
+	// compute remaining rewards for the community pool
+	remaining := consumerRewards.Sub(validatorsRewards)
 
-		k.Logger(ctx).Info(
-			"distributed ICS rewards successfully",
+	// transfer validators rewards to distribution module account
+	validatorsRewardsTrunc, validatorsRewardsChange := validatorsRewards.TruncateDecimal()
+	err = k.bankKeeper.SendCoinsFromModuleToModule(ctx, types.ConsumerRewardsPool, distrtypes.ModuleName, validatorsRewardsTrunc)
+	if err != nil {
+		k.Logger(ctx).Error(
+			"cannot send ICS rewards to distribution module account",
 			"consumerId", consumerId,
 			"chainId", chainId,
-			"total-rewards", consumerRewards.String(),
-			"sent-to-distribution", validatorsRewardsTrunc.String(),
-			"sent-to-CP", remainingRewards.String(),
+			"error", err.Error(),
 		)
+		return types.ConsumerRewardsAllocation{}, err
+	}
+
+	// allocate tokens to consumer validators
+	k.AllocateTokensToConsumerValidators(
+		ctx,
+		consumerId,
+		sdk.NewDecCoinsFromCoins(validatorsRewardsTrunc...),
+	)
 
-		ctx.EventManager().EmitEvent(
-			sdk.NewEvent(
-				types.EventTypeDistributedRewards,
-				sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
-				sdk.NewAttribute(types.AttributeConsumerId, consumerId),
-				sdk.NewAttribute(types.AttributeConsumerChainId, chainId),
-				sdk.NewAttribute(types.AttributeRewardTotal, consumerRewards.String()),
-				sdk.NewAttribute(types.AttributeRewardDistributed, validatorsRewardsTrunc.String()),
-				sdk.NewAttribute(types.AttributeRewardCommunityPool, remainingRewards.String()),
-			),
+	// allocate remaining rewards to the community pool
+	remainingRewards, remainingChanges := remaining.TruncateDecimal()
+	err = k.distributionKeeper.FundCommunityPool(context.Context(ctx), remainingRewards, k.accountKeeper.GetModuleAccount(ctx, types.ConsumerRewardsPool).GetAddress())
+	if err != nil {
+		k.Logger(ctx).Error(
+			"fail to allocate ICS rewards to community pool",
+			"consumerId", consumerId,
+			"chainId", chainId,
+			"error", err.Error(),
 		)
+		return types.ConsumerRewardsAllocation{}, err
+	}
+
+	// set consumer allocations to the remaining rewards decimals
+	alloc.Rewards = validatorsRewardsChange.Add(remainingChanges...)
+
+	k.Logger(ctx).Info(
+		"distributed ICS rewards successfully",
+		"consumerId", consumerId,
+		"chainId", chainId,
+		"total-rewards", consumerRewards.String(),
+		"sent-to-distribution", validatorsRewardsTrunc.String(),
```

Comment:
nit: "sent-to-validators" seems more accurate because the distribution module account includes the CP.